File: BE/CL9/AS2/mpj/test/mpi/perf/Bandwidth.java
package mpi.perf; 
import mpi.*;
import java.nio.ByteBuffer ;
import java.io.* ; 

public class Bandwidth {
  public Bandwidth() {
  }
  public Bandwidth(String[] args) throws Exception {  
    int WARM_UP = 10000 ;
    int REPEAT = 20000 ;
    long[] delays = new long[REPEAT];
    long[] latency = new long[REPEAT];
    long start= 0L, stop=0L, init = 0L;
    MPI.Init(args);				
    int me = MPI.COMM_WORLD.Rank(); 
    System.out.println("Name ="+MPI.Get_processor_name());
    byte byteArray [] = null;
    int j=1, i=0;   	       		   	
    int LOG2N_MAX = 1000000,log2nbyte=0,padding=0;	
    long timed = 0L;		
    byteArray = new byte[16777216];
    PrintStream out = null ;
    FileOutputStream fos = null ;

    
    for(i =0;i < byteArray.length;i++) {
      byteArray[i] = 's';
    }

    /* Can do a warm-up here */
    if(me == 1) {
      fos = new FileOutputStream("bw_delays.out");
      out = new PrintStream (fos);
    }
    else {
      fos = new FileOutputStream("bw.out"); 
      out = new PrintStream (fos);
    } 

    for(i=0;i < 100 ;i++) {
      if(me == 0) {
        MPI.COMM_WORLD.Recv(byteArray,0,1,MPI.BYTE,1,998);
        MPI.COMM_WORLD.Send(byteArray,0,1,MPI.BYTE,1,998);
      }
      else if(me == 1) {
        MPI.COMM_WORLD.Send(byteArray,0,1,MPI.BYTE,0,998);
        MPI.COMM_WORLD.Recv(byteArray,0,1,MPI.BYTE,0,998);
      }
    }

    /* Logrithmic Loop */
    for (log2nbyte = 0; log2nbyte<= LOG2N_MAX && j<16*1024*1024; ++log2nbyte) { 

        //MPI.COMM_WORLD.Barrier();

	j = (1 << log2nbyte);

	/* Latency Calculation Loop */
	for (i = 0; i < REPEAT ; i++) {	   			
	  if(me == 0) {
	    myDelay( (int) (Math.random() * 1000)) ;
	    init = System.nanoTime() ; 
            MPI.COMM_WORLD.Send(byteArray,0,j,MPI.BYTE,1,998);	   		  	    MPI.COMM_WORLD.Recv(byteArray,0,j,MPI.BYTE,1,998);
	    latency[i] = ((System.nanoTime() - init)/1000) ; 
	  } else if(me == 1) {
            MPI.COMM_WORLD.Recv(byteArray,0,j,MPI.BYTE,0,998);
            start = System.nanoTime() ;
	    myDelay( (int)(Math.random() * 1000));
	    delays[i] = (System.nanoTime() - start)/1000 ;
            MPI.COMM_WORLD.Send(byteArray,0,j,MPI.BYTE,0,998);			
	  }
	}

        if(me == 0) {
          for(int k=WARM_UP; k<REPEAT ; k++) {
            out.println(latency[k]+"   ");
          }
        }

        else {
          for (i = WARM_UP ; i < REPEAT ; i++) {
            out.println(delays[i]+"   ");
          }
        }
        
        MPI.COMM_WORLD.Barrier() ;   

	//if(me == 0) { 
        //System.out.println(j+"           "+(latency*1000)+
	//	"           "+(( 8*j ) /( 1024*1024* (latency/(1000)))) );
	//}
    }//end logrithmic loop

    fos.close() ;
    out.close() ;  	 		

    MPI.Finalize();	  	
  }//end args constructor.
  
  static double tripsPerMS = 1000000 ;
  static int dummy ;

  static void myDelay(int us) {

    int trips = (int) (tripsPerMS * us) ;
    long start = System.nanoTime() ;

    for(int i = 0 ; i < trips ; i++) {
      dummy ++ ;
    }

    long actualDelay = System.nanoTime() - start ;

    if(actualDelay > 0 ) {
      long newTripsPerMS = (trips * 1000 ) / actualDelay ;

      if(newTripsPerMS > 0) {
        tripsPerMS = newTripsPerMS ;
      }
    }

  }
 
}//end class.

File: BE/CL9/AS2/mpj/test/mpi/perf/bandwidth.c
#include <stdio.h>
#include "mpi.h"
#define LEN 16777216

int main(int argc, char** argv) 
{  
  MPI_Init(&argc, &argv);				    
  double start=0.0, stop=0.0;
  double totalTime = 0.0;
  //char sbuf[LEN], rbuf[LEN];
  
  MPI_Status status;
  char* sbuf     = (char *) malloc (LEN*sizeof(char));
 // char* rbuf     = (char *) malloc (LEN*sizeof(char));   	
  int WARM_UP = 0;
  int REPEAT = 1;			
  int j=1,i=0;   	       		   	
  int rank =0;
  int LOG2N_MAX = 1000000,log2nbyte=0,padding=0;	
  double timed = 0.0;		
  double latency = 0.0;
    
    for(i =0;i < LEN ;i++) {
      sbuf[i] = 's';
      //rbuf[i] = 'x';
    }
  
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);    

    /* Logrithmic Loop */
    for (log2nbyte = 0; (log2nbyte <= LOG2N_MAX) && (j <= LEN); ++log2nbyte) { 

	j = (1 << log2nbyte);

	/* Warm Up Loop */
	for(i=0;i < WARM_UP ;i++) {
	  if(rank == 0) {	
MPI_Recv(sbuf, j, MPI_CHAR, 1, 998, MPI_COMM_WORLD, &status);		  
MPI_Send(sbuf, j, MPI_CHAR, 1, 998, MPI_COMM_WORLD);		  
	  }
	  else if(rank == 1) {
MPI_Send(sbuf, j, MPI_CHAR, 0, 998, MPI_COMM_WORLD);		  
MPI_Recv(sbuf, j, MPI_CHAR, 0, 998, MPI_COMM_WORLD , &status);		  
	  }		
	}
	
	/* Warm Up Loop */                
	start = MPI_Wtime();				
		
	/* Latency Calculation Loop */
	
	for (i = 0; i < REPEAT ; i++) {	   	
	  if(rank == 0) {
MPI_Send(sbuf, j, MPI_CHAR, 1, 998, MPI_COMM_WORLD);		  
MPI_Recv(sbuf, j, MPI_CHAR, 1, 998, MPI_COMM_WORLD , &status);		  
	  } else if (rank == 1) {
MPI_Recv(sbuf, j, MPI_CHAR, 0, 998, MPI_COMM_WORLD, &status);		  
MPI_Send(sbuf, j, MPI_CHAR, 0, 998, MPI_COMM_WORLD);		  
	  }
	}		
		
	stop = MPI_Wtime();
	timed = stop - start;			
	/* End latency calculation loop */
 	latency = ( ((timed)/(2*REPEAT) )*1000*1000);	
	if(rank == 0) {
        printf("%d\t%.2f\t%.2f\n", j , (latency), 
		            (( 8*j ) /( 1024*1024* (latency/(1000*1000)))) );
	}
    }//end logrithmic loop
    	 		
    MPI_Barrier(MPI_COMM_WORLD);    		
    MPI_Finalize();	  	
    return 0;
}
